sandbox/modules/custom-codecs/src/main/java/org/opensearch/index/codec/customcodec/ZstdNoDictCompressionMode.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.codec.customcodec;
+
+import com.github.luben.zstd.Zstd;
+import java.io.IOException;
+import org.apache.lucene.codecs.compressing.CompressionMode;
+import org.apache.lucene.codecs.compressing.Compressor;
+import org.apache.lucene.codecs.compressing.Decompressor;
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.store.DataOutput;
+import org.apache.lucene.store.ByteBuffersDataInput;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BytesRef;
+
+/** ZSTD Compression Mode (without a dictionary support). */
+public class ZstdNoDictCompressionMode extends CompressionMode {
+
+    private static final int NUM_SUB_BLOCKS = 10;
+    private static final int DEFAULT_COMPRESSION_LEVEL = 6;
+
+    private final int compressionLevel;
+
+    /** default constructor */
+    protected ZstdNoDictCompressionMode() {
+        this.compressionLevel = DEFAULT_COMPRESSION_LEVEL;
+    }
+
+    /**
+     * Creates a new instance with the given compression level.
+     *
+     * @param compressionLevel The compression level.
+     */
+    protected ZstdNoDictCompressionMode(int compressionLevel) {
+        this.compressionLevel = compressionLevel;
+    }
+
+    @Override
+    public Compressor newCompressor() {
+        return new ZSTDCompressor(compressionLevel);
+    }
+
+    @Override
+    public Decompressor newDecompressor() {
+        return new ZSTDDecompressor();
+    }
+
+    /** zstandard compressor */
+    private static final class ZSTDCompressor extends Compressor {
+
+        private final int compressionLevel;
+        private byte[] compressedBuffer;
+
+        /** compressor with a given compresion level */
+        public ZSTDCompressor(int compressionLevel) {
+            this.compressionLevel = compressionLevel;
+            compressedBuffer = BytesRef.EMPTY_BYTES;
+        }
+
+        @Override
+        public void close() throws IOException {}
+
+        private void compress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
+
+            int blockLength = (len + NUM_SUB_BLOCKS - 1) / NUM_SUB_BLOCKS;
+            out.writeVInt(blockLength);
+
+            final int end = off + len;
+
+            for (int start = off; start < end; start += blockLength) {
+                int l = Math.min(blockLength, off + len - start);
```

Comment:
Since all these are signed I would assert to ensure we don't do negative math. That would end up in infinite loops.

Reply:
The latest commit addresses this.